### PR TITLE
:woman_technologist: AVL Tree --- Constructor & `size`

### PR DESCRIPTION
**Pull request by jameshughes89**
### What
Start making an AVL tree class. 

As of now this includes:
- Copy of 102's `BinarySearchTree` interface 
- A `LinkedAvlTree` class with a copy/paste of the nested `Node` class and `size` implemented
- The start of a test class

### Why
It would be one of the topics covered in advanced data structures. Gotta' start somewhere. 

### Testing
:+1: 
Plus, I figure I can start to play with _AssertJ_

### Additional Notes
Guess I have not made the labels for the repo yet. 

I added :mortar_board: and made a new label :woman_technologist: for code content. 

Will rework labels across repos later --- seems like they could be improved. Will do later though, have to go get new windshield wipers.

### CODE REVIEW

File: src/main/java/BinarySearchTree.java [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+
+public interface BinarySearchTree<T extends Comparable<? super T>> extends Iterable<T> {
+
+    /**
+     * Add an element to the tree such that the binary search tree property is maintained. Elements considered equal are
+     * added to the right.
+     *
+     * @param element Element to be added to the tree in the proper location
+     */
+    void add(T element);
```
Comment (twentylemon) [inline src/main/java/BinarySearchTree.java]: We're in second year now, this should probably start returning a boolean. It's typical for mutation operations to return some indication of whether or not they actually changed. That'd also conform with [Collection#add](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/Collection.html#add(E)) then.

File: src/main/java/LinkedAvlTree.java [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+
+public class LinkedAvlTree<T extends Comparable<? super T>> implements BinarySearchTree<T> {
+
+    private int size;
+    private Node<T> root;
+
+    /**
+     * Create a new empty avl tree. The root element is set to null. 
+     */
+    public LinkedAvlTree(){
```
Comment (twentylemon) [inline src/main/java/LinkedAvlTree.java]: Oh formatter, my formatter.

In the tests as well.

**Review by twentylemon — APPROVED** [review]

File: src/main/java/BinarySearchTree.java [hunk 3]
```diff
@@ -6,9 +6,10 @@
      * Add an element to the tree such that the binary search tree property is maintained. Elements considered equal are
      * added to the right.
      *
-     * @param element Element to be added to the tree in the proper location
+     * @param element Element to be added to the tree in the proper location.
+     * @return True if the element was added successfully, False otherwise.
```
Comment (twentylemon) [inline src/main/java/BinarySearchTree.java:10]: `true` and `false`. This is not python.

**Review by twentylemon — APPROVED** [review]

File: src/test/java/LinkedAvlTreeTest.java [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+        @Test
+        void size_empty_returnsZero() {
+            assertThat(classUnderTest.size())
+                    .isEqualTo(0);
+        }
+
+        void isEmpty_empty_returnsTrue() {
+            assertThat(classUnderTest.isEmpty())
+                    .isTrue();
+        }
```
Comment (twentylemon) [inline src/test/java/LinkedAvlTreeTest.java:31]: FYI, these are atypical asserts when you're using assertj. AssertJ provides loads of assertions you can use on the class under test -- you should almost always prefer to use those when possible. For example here,

```java
assertThat(classUnderTest).hasSize(0);
assertThat(classUnderTest).isEmpty();
```

HOWEVER. These assertions could come from [Iterable asserts](https://www.javadoc.io/doc/org.assertj/assertj-core/latest/org/assertj/core/api/AbstractIterableAssert.html#method.summary), which would use the `iterator` to do the work, which you do not have implemented. Further, those asserts would not actually test the `isEmpty` and `size` methods of the class.

TLDR: these asserts are good enough, but ensure you're looking at what assertions are available to do on the actual class under test first. Run tests locally with coverage to confirm their behaviour to yourself, and read a lot of docs. [Core asserts guide](https://assertj.github.io/doc/#assertj-core-assertions-guide)

Comment (jameshughes89) [reply src/test/java/LinkedAvlTreeTest.java:31]: perogies

Comment (jameshughes89) [reply src/test/java/LinkedAvlTreeTest.java:31]: > HOWEVER. These assertions could come from [Iterable asserts](https://www.javadoc.io/doc/org.assertj/assertj-core/latest/org/assertj/core/api/AbstractIterableAssert.html#method.summary), which would use the iterator to do the work, which you do not have implemented. Further, those asserts would not actually test the isEmpty and size methods of the class.

They do, but nevertheless, point stands.
